File: 2023/day-4/part_two.rb
COPIES = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

sum = File.readlines("#{__dir__}/input.txt").sum do |line|
  numbers = line.split(": ").last.split(" | ").map { _1.split.map(&:to_i) }
  matching_numbers = numbers.first & numbers.last

  doubles = COPIES[0]
  COPIES.shift
  COPIES << 1
  matching_numbers.length.times { COPIES[_1] += doubles }

  doubles
end

puts sum
# 6857330

File: 2023/day-4/part_two.ts
import fs from "fs"
import path from "path"

const COPIES = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

fs.readFile(path.resolve(__dirname, "input.txt"), "utf8", (_, lines: string) => {
  const sum = lines.trim().split("\n").reduce((acc, line) => {
    const numbers = line.split(": ").at(-1)!.split(" | ").map(set => set.trim().split(/\s+/).map(number => +number))
    const myNumbers = numbers.at(-1)!

    const matchingNumbers = numbers[0].filter(n => {
      const matchingNumberIndex = myNumbers.findIndex(m => m === n)
      if (matchingNumberIndex === -1) { return false }

      myNumbers.splice(matchingNumberIndex, 1)
      return true
    })

    const doubles = COPIES[0]
    COPIES.splice(0, 1)
    COPIES.push(1)
    matchingNumbers.forEach((_, index) => { COPIES[index] += doubles })

    return acc + doubles
  }, 0)

  console.log(sum)
  // 6857330
})
